src/daft-io/src/object_io.rs

```diff
@@ -112,10 +118,240 @@ pub(crate) trait ObjectSource: Sync + Send {
+    source: Arc<dyn ObjectSource>,
+    glob: &str,
+) -> super::Result<BoxStream<'static, super::Result<FileMetadata>>> {
+    // If no special characters, we fall back to ls behavior
+    let full_fragment = GlobFragment::new(glob);
+    if !full_fragment.has_special_character() {
+        let glob = full_fragment.escaped_str().to_string();
+        return Ok(stream! {
+            let mut results = source.iter_dir(glob.as_str(), Some("/"), None).await?;
+            while let Some(val) = results.next().await {
+                match &val {
+                    // Ignore non-File results
+                    Ok(fm) if !matches!(fm.filetype, FileType::File) => continue,
+                    _ => yield val,
+                }
+            }
+        }
+        .boxed());
+    }
+
+    // If user specifies a trailing / then we understand it as an attempt to list the folder(s) matched
+    // and append a trailing * fragment
+    let glob = if glob.ends_with('/') {
+        glob.to_string() + "*"
+    } else {
+        glob.to_string()
+    };
+    let glob = glob.as_str();
+
+    let glob_fragments = to_glob_fragments(glob)?;
+    let full_glob_matcher = GlobBuilder::new(glob)
+        .literal_separator(true)
+        .backslash_escape(true)
+        .build()
+        .map_err(|err| super::Error::InvalidArgument {
+            msg: format!("Cannot parse provided glob {glob}: {err}"),
+        })?
+        .compile_matcher();
+
+    // Channel to send results back to caller. Note that all results must have FileType::File.
+    let (to_rtn_tx, mut to_rtn_rx) = tokio::sync::mpsc::channel(16 * 1024);
+
+    /// Dispatches a task to visit the specified `path` (a concrete path on the filesystem to either a File or Directory).
+    /// Based on the current glob_fragment being processed (accessible via `glob_fragments[i]`) this task will:
+    ///   1. Perform work to retrieve Files/Directories at (`path` + `glob_fragments[i]`)
+    ///   2. Return results to the provided `result_tx` channel based on the provided glob, if appropriate
+    ///   3. Dispatch additional tasks via `.visit()` to continue visiting them, if appropriate
+    fn visit(
+        result_tx: Sender<super::Result<FileMetadata>>,
+        source: Arc<dyn ObjectSource>,
+        state: GlobState,
+    ) {
+        tokio::spawn(async move {
+            log::debug!(target: "glob", "Visiting '{}' with glob_fragments: {:?}", &state.current_path, &state.glob_fragments);
+            let current_fragment = state.current_glob_fragment();
+
+            // BASE CASE: current_fragment is a **
+            // We perform a recursive ls and filter on the results for only FileType::File results that match the full glob
+            if current_fragment.escaped_str() == "**" {
+                let mut results = source
+                    .iter_dir(&state.current_path, Some("/"), None)
+                    .await
+                    .unwrap_or_else(|e| stream! {yield Err(e)}.boxed());
```

Comment:
would avoid the `stream!` macro here as well. You should be able to do `futures::stream::iter([Err(e)])`